pennylane/operation.py

```diff
@@ -713,6 +713,17 @@ def inverse(self):
         """Boolean determining if the inverse of the operation was requested."""
         return self._inverse
 
+    def adjoint(self, do_queue=False):
```

Comment:
Is this an optional abstract method for operations to define?

Reply:
For now yes. When I added a abc.abstractmethod decorator, the photonic ops (beam splitter, etc) would complain that they didn't have an adjoint method. My photonics isn't strong enough to know what an adjoint of a beamsplitter should be, so I left it as an optional interface.

Reply:
Ah, it's pretty simple for the photonic gates; typically we just negate the first parameter (since they are of the form `exp(-itK)`:

* Beamsplitter: negate first parameter
* ControlledAddition: negate parameter
* ControlledPhase: negate parameter
* Displacement: negate parameter
* Kerr: negate parameter
* CrossKerr: negate parameter
* QuadraticPhase: negate parameter
* Rotation: negate parameter
* Squeezing: negate first parameter
* TwoModeSqueezing: negate first parameter
* CubicPhase: negate parameter
* Interferometer: conjugate transpose parameter

For state preparations, I'm not sure the adjoint is well defined:

* CatState: not applicable
* CoherentState: not applicable
* FockDensityMatrix: not applicable
* DisplacedSqueezedState: not applicable
* FockState: not applicable
* FockStateVector: not applicable
* SqueezedState: not applicable
* ThermalState: not applicable
* GaussianState: not applicable (transpose?)

Reply:
I'll add most of these in a next PR to keep this one small.